jgrapht-core/src/main/java/org/jgrapht/traverse/DepthFirstIterator.java

```diff
@@ -34,42 +34,14 @@
  * @param <V> the graph vertex type
  * @param <E> the graph edge type
  *
- * @author Liviu Rau
- * @author Barak Naveh
- * @since Jul 29, 2003
+ * @author Joris Kinable
  */
 public class DepthFirstIterator<V, E>
-    extends CrossComponentIterator<V, E, DepthFirstIterator.VisitColor>
+    extends CrossComponentIterator<V, E, DepthFirstIterator.SearchNodeData>
 {
-    /**
-     * Sentinel object. Unfortunately, we can't use null, because ArrayDeque won't accept those. And
-     * we don't want to rely on the caller to provide a sentinel object for us. So we have to play
-     * typecasting games.
-     */
-    public static final Object SENTINEL = new Object();
-
-    /**
-     * Standard vertex visit state enumeration.
-     */
-    protected static enum VisitColor
-    {
-        /**
-         * Vertex has not been returned via iterator yet.
-         */
-        WHITE,
-
-        /**
-         * Vertex has been returned via iterator, but we're not done with all of its out-edges yet.
-         */
-        GRAY,
-
-        /**
-         * Vertex has been returned via iterator, and we're done with all of its out-edges.
-         */
-        BLACK
-    }
 
-    private Deque<Object> stack = new ArrayDeque<>();
+    //private Stack<V> stack = new Stack<>();
+    private Stack<StackEntry> stack = new Stack<>();
```

Comment:
You seem to raise a good point. I completely forgot about this. I always find this amusingly contradictory: "This class is likely to be faster than Stack when used as a stack".
Any idea whether it is better to use a ArrayDeque or a LinkedList as stack?